### PR TITLE
Help inference inside cat related functions

### PR DESCRIPTION
**Pull request by p-zubieta**
This is an easy fix for the problem mentioned in https://github.com/JuliaLang/julia/issues/16128#issuecomment-224633749 only for cat related functions. It won't help in general for similar situations but I leave it here in case people decide this is worth anyway.

CC @JeffBezanson

### CODE REVIEW

File: base/abstractarray.jl [hunk 1]
```diff
@@ -613,8 +613,8 @@ cat(catdim::Integer) = Array{Any}(0)
 
 vcat() = Array{Any}(0)
 hcat() = Array{Any}(0)
-typed_vcat(T::Type) = Array{T}(0)
-typed_hcat(T::Type) = Array{T}(0)
+typed_vcat{T}(TT::Type{T}) = Array{T}(0)
+typed_hcat{T}(TT::Type{T}) = Array{T}(0)
```
Comment (JeffBezanson) [inline base/abstractarray.jl]: You can leave the argument name out in this case; `::Type{T}`.

Comment (pabloferz) [inline base/abstractarray.jl]: I forgot about this for a moment. I'll change it.

**Comment by pabloferz** [comment]
Thinking about this a bit more. Should this also be done all over base irrespectively of whether or not inference can be made to work again in this cases?

**Comment by tkelman** [comment]
It's kinda messy, if it helps measurably then I guess it's worth applying the workaround for now, but hopefully whatever is preventing it from working can be fixable?

**Comment by JeffBezanson** [comment]
I don't think we should apply this change in every case in Base; it's overkill. Let's just do it here and probably also for `similar`, for now.

**Comment by pabloferz** [comment]
You're right, it would be too much. @tkelman It definitively improves things, but I'll check what is the minimum that need to be changed so it helps without increasing too much the number of compiled methods.

**Pull request merged**